### PR TITLE
Issue #1560: improved help and debugging info when a command fails

### PR DESCRIPTION
**Pull request by greggles**
For https://github.com/drush-ops/drush/issues/1560

I didn't make a branch for this - sorry :/ I can do that if you'd like.

### CODE REVIEW

**Comment by jonhattan** [comment]
The test fail is unrelated. The cause is libraries module with latest core https://www.drupal.org/node/2557419

File: includes/backend.inc [hunk 1]
```diff
@@ -1022,7 +1022,8 @@ function _drush_backend_invoke($cmds, $common_backend_options = array(), $contex
     $ret = array();
     foreach ($procs as $site => $proc) {
       if (($proc['code'] == DRUSH_APPLICATION_ERROR) && isset($common_backend_options['integrate'])) {
-        drush_set_error('DRUSH_APPLICATION_ERROR', dt("The external command could not be executed due to an application error."));
+        drush_set_error('DRUSH_APPLICATION_ERROR', dt("The external command could not be executed due to an application error. Try running the command again with the --debug option."));
+        drush_log(dt('The raw command and output is !proc', array('!proc' => var_export($proc, TRUE))), 'debug');
```
Comment (weitzman) [inline includes/backend.inc]: 1. $proc can be giant, with log messages, return value, errors, and output. I think this should be $proc['output']? Paging @greg-1-anderson.
1. Furthermore, there is a line later which attempts to log the output. Would be good to understand why that didn't help you. See `$ret = drush_set_error('DRUSH_FRAMEWORK_ERROR', dt("The command could not be executed successfully (returned: !return, code: !code)", array("!return" => $proc['output'], "!code" =>  $proc['code'])));`

**Comment by greg-1-anderson** [comment]
Could you provide the exact situation you encountered that prompted you to submit this?

Note that in the case of a DRUSH_APPLICATION_ERROR, this usually indicates that Drush itself did not start at all, so $proc might be smaller than usual inside this particular conditional.  Not sure that this is always the case, though, and in any event, as Moshe said, `$proc['output']` should have the info you need already.  Let's chase this down before deciding what to do here.

File: includes/backend.inc (hunk 1 above)
Comment (greggles) [inline includes/backend.inc]: I debated about whether to dump all of $proc or not. I'd be fine with just output, but I feel like using --debug is asking for "all the data".

**Comment by greggles** [comment]
I can describe the situation, but I'm not sure it will be reproducible on other systems. I can reproduce it locally and in the test environment (this runs as part of a script in test for me).

* sites/default/settings.php had a fatal error in it - require_once of a nonexistent file.
* the alias I was using set keys for root and uri that should not have loaded sites/default/settings.php and instead should have loaded sites/example.backup/settings.php
* I was running `sql-sanitize -y` and oddly enough that command succeeded...but right before it printed the list of things it was going to do...it printed this error message and nothing else
* I was running the command from outside of the docroot

Here's a gist of the debug output from my local environment https://gist.github.com/greggles/2fbb8464cb7eacf63bda

The error is on line 21 and there's no additional output.

I think it makes sense to add the sentence about --debug regardless of outputting `$proc`.

**Comment by weitzman** [comment]
This bug report applies to Drush6. I'll leave it to other interested parties to determine the value of this PR. Feel free to reopen if new info arises.

**Comment by greggles** [comment]
My memory of why I closed this was that we discussed it in irc (or similar) and the feeling by drush maintainers that it wasn't appropriate.

I just helped someone in the #drush slackroom who had a silent fail of drush. They had to go look in syslog to find the problem.

It still seems odd to me that drush silently hides errors instead of showing it to people.

**Pull request closed without merging**